### PR TITLE
feat(auth-provider): add support for oidc auth provider

### LINKED ISSUE #218: Support OpenID Connect standard
Would be a great addition to this lib to support OpenID Connect Oauth standard, to configure and provide custom auth providers.

### LINKED ISSUE #253: Support Keycloak Oauth

### PR DESCRIPTION
**Pull request by azlekov**
closes #253
closes #218

References:
- https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims
- https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
- https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse
- https://auth0.com/docs/get-started/apis/scopes/openid-connect-scopes
- https://auth0.com/docs/get-started/apis/scopes/openid-connect-scopes#standard-claims

Before submitting this PR:

### Checklist

- [x] No breaking changes
- [ ] Tests pass
- [ ] New features have new tests
- [x] Documentation is updated

### Breaking changes

Avoid breaking changes and regressions. If you feel it is unavoidable, make it explicit in your PR comment so we can review it and see how to handle it.

### Tests

- please make sure your changes pass the current tests (Use the `make test` or the `make watch` command).
- if you are introducing a new feature, please write as much tests as possible.

### Documentation

Please make sure the documentation is updated accordingly, in particular:

- [Workflows](https://github.com/nhost/hasura-auth/tree/main/docs/workflows). Workflows are [Mermaid sequence diagrams](https://mermaid-js.github.io/mermaid/#/sequenceDiagram)
- [Schema](https://github.com/nhost/hasura-auth/blob/main/docs/schema.md). The schema in a [Mermaid ER diagram](https://mermaid-js.github.io/mermaid/#/entityRelationshipDiagram)
- [Environment variables](https://github.com/nhost/hasura-auth/blob/main/docs/environment-variables.md). Please adjust the [.env.example](https://github.com/nhost/hasura-auth/blob/main/.env.example) file accordingly
- OpenApi specifications. We are using inline [JSDoc annotations](https://www.npmjs.com/package/express-jsdoc-swagger)

### CODE REVIEW

**Comment by azlekov** [comment]
@plmercereau can you help me with failing pipeline...

**Comment by plmercereau** [comment]
done
that's an annoying timeout in downloading the nhost cli from GitHub...
Eventually we should create an action that caches the cli, if you have time to spare..

**Comment by azlekov** [comment]
Tested with Keycloak. Supports basic and common flows and configurations.

The authorization URL differs between different OIDC providers, some ends with `/authorize`, other with `/auth`, another with `/athorization` etc. that's why I made separate env variables for all interesting endpoints. This way more OIDC providers, should be covered out of the box.

@plmercereau can you take a look, it's rather simple thanks to your quite nice and comprehensive grant integration.

File: src/routes/oauth/config.ts [hunk 1]
```diff
@@ -385,4 +385,25 @@ export const PROVIDERS_CONFIG: Record<
       next();
     },
   },
+  oidc: {
+    grant: {
+      oauth: 2,
+      nonce: true,
+      scope_delimiter: ' ',
+      scope: ['openid', 'profile', 'email'],
```
Comment (elitan) [inline src/routes/oauth/config.ts:393]: Do all OIDC providers have these three scopes? Are there some official OIDC documentation around this?

File: src/routes/oauth/config.ts [hunk 2]
```diff
@@ -385,4 +385,25 @@ export const PROVIDERS_CONFIG: Record<
...
+      profile_url: `${process.env.AUTH_PROVIDER_OIDC_USER_INFO_URL}`,
+      client_id: process.env.AUTH_PROVIDER_OIDC_CLIEND_ID,
+      client_secret: process.env.AUTH_PROVIDER_OIDC_CLIENT_SECRET,
+    },
+    profile: ({ profile }) => ({
+      id: profile.sub,
+      email: profile.email,
+      emailVerified: profile.email_verified,
+      displayName: profile.name && profile.nickname,
+      avatarUrl: profile.picture,
```
Comment (elitan) [inline src/routes/oauth/config.ts:406]: Similar question here. Is this official OIDC profile properties?

````
- email
- email_verified
- name
- nickname
- picture
```

File: src/routes/oauth/config.ts (hunk 1 above)
Comment (azlekov) [reply src/routes/oauth/config.ts:393]: Yes:
- https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims
- https://auth0.com/docs/get-started/apis/scopes/openid-connect-scopes
- https://learn.microsoft.com/EN-US/azure/active-directory/develop/scopes-oidc

File: src/routes/oauth/config.ts (hunk 2 above)
Comment (azlekov) [reply src/routes/oauth/config.ts:406]: Yes, but looks like not all providers returns all of them, that's why I'm trying fallback in some cases (for example with the displayName)

- https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
- https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse
- https://auth0.com/docs/get-started/apis/scopes/openid-connect-scopes#standard-claims

**Comment by azlekov** [comment]
Hey @elitan can you check why the pipeline is failing? Do I need to do something further to get this merged?

**Comment by remaininlight** [comment]
This PR would be awesome to have in NHost :) I can't work out why the signout test in the pipeline is failing but @elitan if you have time it would be much appreciated

**Comment by remaininlight** [comment]
Hi @szilarddoro & @elitan,

Is there any update on this please? Would be very useful addition to NHost, seems like there is a fair amount of interest

Hope you're both well!

**Pull request closed without merging**